File: node/test/test_bindings.js
'use strict';
/**
 * node/test/test_bindings.js — smoke-test all major Node.js binding surfaces.
 *
 * Run from repo root:
 *   cd node && node test/test_bindings.js
 */

const path = require('path');
const fs   = require('fs');
const flox = require(path.join(__dirname, '..'));

let passed = 0;
let failed = 0;

function check(condition, msg) {
  if (condition) {
    console.log(`  ok  ${msg}`);
    passed++;
  } else {
    console.error(`  FAIL  ${msg}`);
    failed++;
  }
}

function approx(a, b, eps = 1e-6) {
  return isFinite(a) && isFinite(b) && Math.abs(a - b) < eps;
}

// ── Constants ─────────────────────────────────────────────────────────

console.log('=== Constants ===');
check(flox.POSITION_FIFO === 0, 'POSITION_FIFO === 0');
check(flox.POSITION_AVG_COST === 1, 'POSITION_AVG_COST === 1');
check(flox.SLIPPAGE_NONE === 0, 'SLIPPAGE_NONE === 0');
check(flox.SLIPPAGE_FIXED_TICKS === 1, 'SLIPPAGE_FIXED_TICKS === 1');
check(flox.SLIPPAGE_FIXED_BPS === 2, 'SLIPPAGE_FIXED_BPS === 2');
check(flox.SLIPPAGE_VOLUME_IMPACT === 3, 'SLIPPAGE_VOLUME_IMPACT === 3');
check(flox.QUEUE_NONE === 0, 'QUEUE_NONE === 0');
check(flox.QUEUE_TOB === 1, 'QUEUE_TOB === 1');
check(flox.QUEUE_FULL === 2, 'QUEUE_FULL === 2');

// ── Streaming indicators ──────────────────────────────────────────────

console.log('=== Streaming indicators ===');

const sma = new flox.SMA(3);
sma.update(1.0); sma.update(2.0); sma.update(3.0);
check(approx(sma.value, 2.0), 'SMA(3) of [1,2,3] == 2.0');
sma.update(4.0);
check(approx(sma.value, 3.0), 'SMA(3) shifts to [2,3,4] == 3.0');
check(sma.ready === true, 'SMA.ready === true');

const ema = new flox.EMA(3);
for (let i = 1; i <= 10; i++) ema.update(i);
check(ema.value > 0, 'EMA.value > 0 after 10 updates');
check(ema.ready === true, 'EMA.ready === true');

const rsi = new flox.RSI(14);
for (let i = 1; i <= 20; i++) rsi.update(i);
check(rsi.value >= 0 && rsi.value <= 100, `RSI in [0,100], got ${rsi.value.toFixed(2)}`);

const atr = new flox.ATR(14);
for (let i = 0; i < 20; i++) atr.update(100 + i, 99 + i, 100 + i);
check(atr.value > 0, 'ATR > 0');

// ── Batch indicators ──────────────────────────────────────────────────

console.log('=== Batch indicators ===');

const prices = new Float64Array([1, 2, 3, 4, 5, 6, 7]);
const bSma = flox.sma(prices, 3);
check(approx(bSma[2], 2.0), 'batch sma[2] == 2.0');
check(approx(bSma[6], 6.0), 'batch sma[6] == 6.0');

const bEma = flox.ema(prices, 3);
check(bEma.length === prices.length, 'batch ema length matches input');
check(bEma[6] > bEma[2], 'batch ema increases for ascending input');

// ── Targets (forward-looking labels) ──────────────────────────────────

console.log('=== Targets ===');

const tClose = new Float64Array([100.0, 101.0, 99.0, 105.0, 110.0]);
const fr = flox.targets.future_return(tClose, 2);
check(approx(fr[0], 99.0 / 100.0 - 1.0), 'targets.future_return[0]');
check(approx(fr[2], 110.0 / 99.0 - 1.0), 'targets.future_return[2]');
check(Number.isNaN(fr[3]) && Number.isNaN(fr[4]), 'targets.future_return tail NaN');

const constClose = new Float64Array(20).fill(100.0);
const vol = flox.targets.future_ctc_volatility(constClose, 5);
check(approx(vol[0], 0.0), 'targets.future_ctc_volatility const-series == 0');
check(Number.isNaN(vol[19]), 'targets.future_ctc_volatility tail NaN');

const linear = new Float64Array(20);
for (let i = 0; i < 20; ++i) linear[i] = 100.0 + 0.5 * i;
const sl = flox.targets.future_linear_slope(linear, 4);
check(approx(sl[0], 0.5), 'targets.future_linear_slope linear-series == 0.5');
check(Number.isNaN(sl[19]), 'targets.future_linear_slope tail NaN');

// ── IndicatorGraph (batch) ────────────────────────────────────────────

console.log('=== IndicatorGraph (batch) ===');

const ramp = new Float64Array(50);
for (let i = 0; i < 50; ++i) ramp[i] = i;

const g = new flox.IndicatorGraph();
g.setBars(0, ramp);

g.addNode('ema5', [], (graph, sym) => flox.ema(graph.close(sym), 5));
g.addNode('sma5', [], (graph, sym) => flox.sma(graph.close(sym), 5));
g.addNode('diff', ['ema5', 'sma5'], (graph, sym) => {
  const a = graph.get(sym, 'ema5');
  const b = graph.get(sym, 'sma5');
  const out = new Float64Array(a.length);
  for (let i = 0; i < a.length; ++i) out[i] = a[i] - b[i];
  return out;
});

const ema5 = g.require(0, 'ema5');
const diff = g.require(0, 'diff');
check(ema5.length === 50, 'graph require returns full-length array');
check(diff.length === 50, 'graph dependent node has same length');
check(g.get(0, 'sma5') !== null, 'graph get returns cached node array');
check(g.get(0, 'never_added') === null, 'graph get on missing node returns null');

let threw = false;
try { g.require(0, 'missing'); } catch (e) { threw = true; }
check(threw, 'graph require on unknown node throws');

// ── PositionTracker ───────────────────────────────────────────────────

console.log('=== PositionTracker ===');

const pt = new flox.PositionTracker(flox.POSITION_FIFO);
pt.onFill(0, 'buy', 50000.0, 1.0);
check(approx(pt.position(0), 1.0), 'position == 1.0 after buy');
check(approx(pt.avgEntryPrice(0), 50000.0), 'avgEntryPrice == 50000');
check(approx(pt.realizedPnl(0), 0.0), 'realizedPnl == 0 before close');

pt.onFill(0, 'sell', 51000.0, 1.0);
check(approx(pt.position(0), 0.0), 'position == 0 after sell');
check(pt.realizedPnl(0) > 0, 'realizedPnl > 0 on winning trade');
check(approx(pt.realizedPnl(0), 1000.0), 'realizedPnl == 1000');
check(approx(pt.totalRealizedPnl(), 1000.0), 'totalRealizedPnl == 1000');

// Multi-symbol
const pt2 = new flox.PositionTracker(flox.POSITION_FIFO);
pt2.onFill(0, 'buy', 100.0, 2.0);
pt2.onFill(1, 'sell', 200.0, 3.0);
check(approx(pt2.position(0), 2.0), 'symbol 0 position == 2.0');
check(approx(pt2.position(1), -3.0), 'symbol 1 position == -3.0');

// ── OrderTracker ──────────────────────────────────────────────────────

console.log('=== OrderTracker ===');

const ot = new flox.OrderTracker();
check(ot.onSubmitted(1001, 0, 'buy', 50000.0, 1.0) === true, 'onSubmitted returns true');
check(ot.isActive(1001) === true, 'order 1001 is active');
check(ot.activeCount === 1, 'activeCount === 1');
check(ot.totalCount === 1, 'totalCount === 1');

ot.onSubmitted(1002, 0, 'sell', 51000.0, 1.0);
check(ot.activeCount === 2, 'activeCount === 2 after second order');

check(ot.onFilled(1001, 1.0) === true, 'onFilled returns true');
check(ot.isActive(1001) === false, 'order 1001 inactive after fill');
check(ot.activeCount === 1, 'activeCount === 1 after fill');

check(ot.onCanceled(1002) === true, 'onCanceled returns true');
check(ot.isActive(1002) === false, 'order 1002 inactive after cancel');
check(ot.activeCount === 0, 'activeCount === 0');
check(ot.totalCount === 2, 'totalCount === 2');

// ── Engine (data loading) ─────────────────────────────────────────────

console.log('=== Engine ===');

const csvPath = path.join(__dirname, '..', '..', 'docs', 'examples', 'data', 'btcusdt_1m.csv');
if (fs.existsSync(csvPath)) {
  const engine = new flox.Engine();
  engine.loadCsv(csvPath);
  const n = engine.barCount();
  check(n > 1000, `barCount ${n} > 1000`);
  const opens  = engine.open();
  const closes = engine.close();
  const highs  = engine.high();
  const lows   = engine.low();
  check(opens.length  === n, 'opens.length  == barCount');
  check(closes.length === n, 'closes.length == barCount');
  check(opens[0] > 0, 'first open price > 0');
  check(highs.every((h, i) => h >= lows[i]), 'high >= low for every bar');
  console.log(`  loaded ${n} bars, price ${Math.min(...closes).toFixed(0)}–${Math.max(...closes).toFixed(0)}`);
} else {
  console.log(`  skip Engine (CSV not found: ${csvPath})`);
}

// ── Summary ───────────────────────────────────────────────────────────

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: python/tests/test_bindings.py
"""
python/tests/test_bindings.py — smoke-test all major Python binding surfaces.

Run from repo root:
    PYTHONPATH=build/python python3 python/tests/test_bindings.py
"""

import sys
import os
import math

build_dir = os.path.join(os.path.dirname(__file__), '..', '..', 'build', 'python')
sys.path.insert(0, os.path.abspath(build_dir))

import flox_py as flox
import numpy as np

_passed = 0
_failed = 0


def check(condition, msg):
    global _passed, _failed
    if condition:
        print(f"  ok  {msg}")
        _passed += 1
    else:
        print(f"  FAIL  {msg}")
        _failed += 1


def approx(a, b, eps=1e-6):
    return math.isfinite(a) and math.isfinite(b) and abs(a - b) < eps


# ── Constants ────────────────────────────────────────────────────────

print("=== Constants ===")

check(flox.SLIPPAGE_NONE == 0, "SLIPPAGE_NONE == 0")
check(flox.SLIPPAGE_FIXED_TICKS == 1, "SLIPPAGE_FIXED_TICKS == 1")
check(flox.SLIPPAGE_FIXED_BPS == 2, "SLIPPAGE_FIXED_BPS == 2")
check(flox.SLIPPAGE_VOLUME_IMPACT == 3, "SLIPPAGE_VOLUME_IMPACT == 3")
check(flox.QUEUE_NONE == 0, "QUEUE_NONE == 0")
check(flox.QUEUE_TOB == 1, "QUEUE_TOB == 1")
check(flox.QUEUE_FULL == 2, "QUEUE_FULL == 2")

check(flox.PRICE_SCALE == 100_000_000, "PRICE_SCALE == 1e8")
check(flox.QUANTITY_SCALE == 100_000_000, "QUANTITY_SCALE == 1e8")
check(flox.VOLUME_SCALE == 100_000_000, "VOLUME_SCALE == 1e8")

# ── Raw-to-double converters ─────────────────────────────────────────

print("=== Raw-to-double converters ===")

# Contiguous int64 input
contig = np.array([100_00000000, 200_00000000, 300_00000000], dtype=np.int64)
out = flox.prices_to_double(contig)
check(np.allclose(out, [100.0, 200.0, 300.0]), "prices_to_double on contiguous 1D")
check(out.dtype == np.float64, "prices_to_double returns float64")

# Non-contiguous slice (every-other element)
strided_src = np.array([1, 999, 2, 999, 3, 999], dtype=np.int64) * 100_000_000
strided = strided_src[::2]
check(not strided.flags["C_CONTIGUOUS"], "test setup: strided view is non-contiguous")
out = flox.prices_to_double(strided)
check(np.allclose(out, [1.0, 2.0, 3.0]),
      "prices_to_double on non-contiguous strided view (every-other element)")

# Field of a structured array — the canonical buggy case
bar_dtype = np.dtype([
    ("start_time_ns", "<i8"),
    ("end_time_ns", "<i8"),
    ("open_raw", "<i8"),
    ("high_raw", "<i8"),
    ("low_raw", "<i8"),
    ("close_raw", "<i8"),
    ("volume_raw", "<i8"),
    ("buy_volume_raw", "<i8"),
    ("trade_count", "<i8"),
])
bars = np.zeros(5, dtype=bar_dtype)
bars["close_raw"] = np.array([100_00000000, 101_00000000, 102_00000000,
                              103_00000000, 104_00000000], dtype=np.int64)
# Set other fields to garbage to ensure we don't accidentally read them
bars["open_raw"] = 999_00000000
bars["volume_raw"] = 7777_00000000
out = flox.prices_to_double(bars["close_raw"])
check(np.allclose(out, [100.0, 101.0, 102.0, 103.0, 104.0]),
      "prices_to_double on structured-array field view (close_raw)")

out_v = flox.volumes_to_double(bars["volume_raw"])
check(np.allclose(out_v, [7777.0] * 5), "volumes_to_double on structured-array field view")

# 2D input — shape preserved
two_d = np.arange(6, dtype=np.int64).reshape(2, 3) * 100_000_000
out2 = flox.quantities_to_double(two_d)
check(out2.shape == (2, 3), "quantities_to_double preserves 2D shape")
check(np.allclose(out2, [[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]]), "quantities_to_double 2D values")

# ── Streaming indicators ──────────────────────────────────────────────

print("=== Streaming indicators ===")

sma = flox.SMA(3)
sma.update(1.0); sma.update(2.0); sma.update(3.0)
check(approx(sma.value, 2.0), "SMA(3) of [1,2,3] == 2.0")
sma.update(4.0)
check(approx(sma.value, 3.0), "SMA(3) shifts to [2,3,4] == 3.0")

ema = flox.EMA(3)
for v in range(1, 11):
    ema.update(float(v))
check(ema.value > 0, "EMA value > 0 after 10 updates")
check(ema.value > 5.0, "EMA(10 ascending) > 5.0")

rsi = flox.RSI(14)
for v in range(1, 30):
    rsi.update(float(v))
check(0.0 <= rsi.value <= 100.0, f"RSI in [0,100], got {rsi.value:.2f}")

atr = flox.ATR(14)
for i in range(20):
    atr.update(100.0 + i, 99.0 + i, 100.0 + i)
check(atr.value > 0.0, "ATR > 0")

# MACD
macd = flox.MACD(3, 6, 2)
for v in range(1, 20):
    macd.update(float(v))
check(math.isfinite(macd.value), "MACD.value is finite")

# Bollinger
bb = flox.Bollinger(5, 2.0)
for v in [10.0, 11.0, 12.0, 10.0, 11.0, 12.0, 13.0]:
    bb.update(v)
check(math.isfinite(bb.value), "Bollinger.value is finite")

# ── Batch indicators ──────────────────────────────────────────────────

print("=== Batch indicators ===")

prices = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0])
out = flox.sma(prices, 3)
check(approx(out[2], 2.0), "batch sma[2] == 2.0")
check(approx(out[6], 6.0), "batch sma[6] == 6.0")

out_ema = flox.ema(prices, 3)
check(len(out_ema) == len(prices), "batch ema length matches input")
check(out_ema[6] > out_ema[2], "batch ema is increasing for ascending input")

# ── AutoCorrelation ──────────────────────────────────────────────────

print("=== AutoCorrelation ===")

# Linear y = a + b*t -> lag-1 autocorrelation == 1.
linear = np.array([5.0 + 0.7 * i for i in range(50)])
ac = flox.autocorrelation(linear, 10, 1)
check(math.isnan(ac[9]), "autocorrelation warmup at index < window+lag-1 is NaN")
check(approx(ac[10], 1.0), "autocorrelation on linear series, lag=1 == 1.0")
check(approx(ac[40], 1.0), "autocorrelation on linear series at later index == 1.0")

# Streaming class matches the batch function on the same input.
stream = flox.AutoCorrelation(10, 1)
streamed = []
for v in linear:
    streamed.append(stream.update(float(v)))
# Streaming is ready at index window+lag-1 = 10 (0-based).
for i in range(10, len(linear)):
    check(approx(float(streamed[i]), float(ac[i])),
          f"streaming AC matches batch at i={i}")
    if i >= 10:
        break  # one assertion is enough to keep noise down

# ── PositionTracker ───────────────────────────────────────────────────

print("=== PositionTracker ===")

pt = flox.PositionTracker("fifo")
pt.on_fill(0, "buy", 50000.0, 1.0)
check(approx(pt.position(0), 1.0), "position == 1.0 after buy")
check(approx(pt.avg_entry_price(0), 50000.0), "avg_entry_price == 50000")
check(approx(pt.realized_pnl(0), 0.0), "realized_pnl == 0 before close")

pt.on_fill(0, "sell", 51000.0, 1.0)
check(approx(pt.position(0), 0.0), "position == 0 after sell")
check(pt.realized_pnl(0) > 0.0, "realized_pnl > 0 on winning trade")
check(approx(pt.realized_pnl(0), 1000.0), "realized_pnl == 1000")
check(approx(pt.total_realized_pnl(), 1000.0), "total_realized_pnl == 1000")

# Average cost method
pt_avg = flox.PositionTracker("average")
pt_avg.on_fill(0, "buy", 100.0, 2.0)
pt_avg.on_fill(0, "buy", 200.0, 2.0)
check(approx(pt_avg.position(0), 4.0), "avg cost: position == 4.0")
check(approx(pt_avg.avg_entry_price(0), 150.0), "avg cost: avg_entry == 150.0")

# Multi-symbol
pt2 = flox.PositionTracker("fifo")
pt2.on_fill(0, "buy", 100.0, 2.0)
pt2.on_fill(1, "sell", 200.0, 3.0)
check(approx(pt2.position(0), 2.0), "symbol 0 position == 2.0")
check(approx(pt2.position(1), -3.0), "symbol 1 position == -3.0")

# ── OrderTracker ─────────────────────────────────────────────────────

print("=== OrderTracker ===")

ot = flox.OrderTracker()
check(ot.on_submitted(1001, "EX-001") == True, "on_submitted returns True")
check(ot.is_active(1001), "order 1001 is active")
check(ot.active_count() == 1, "active_count == 1")
check(ot.total_count() == 1, "total_count == 1")

ot.on_submitted(1002, "EX-002")
check(ot.active_count() == 2, "active_count == 2 after second order")

check(ot.on_filled(1001, 1.0) == True, "on_filled returns True")
check(not ot.is_active(1001), "order 1001 inactive after fill")
check(ot.active_count() == 1, "active_count == 1 after fill")

check(ot.on_canceled(1002) == True, "on_canceled returns True")
check(not ot.is_active(1002), "order 1002 inactive after cancel")
check(ot.active_count() == 0, "active_count == 0")
check(ot.total_count() == 2, "total_count == 2")

# ── Engine (data loading) ────────────────────────────────────────────

print("=== Engine ===")

csv_path = os.path.abspath(
    os.path.join(os.path.dirname(__file__), '..', '..', 'docs', 'examples', 'data', 'btcusdt_1m.csv')
)
if os.path.exists(csv_path):
    engine = flox.Engine()
    engine.load_csv(csv_path)
    n = engine.bar_count()
    check(n > 1000, f"bar_count {n} > 1000")
    opens = engine.open()
    closes = engine.close()
    highs = engine.high()
    lows = engine.low()
    check(len(opens) == n, "len(opens) == bar_count")
    check(len(closes) == n, "len(closes) == bar_count")
    check(all(opens[:10] > 0), "first 10 open prices > 0")
    check(all(highs >= lows), "high >= low for every bar")
    print(f"  loaded {n} bars, price {min(closes):.0f}–{max(closes):.0f}")
else:
    print(f"  skip Engine (CSV not found: {csv_path})")

# ── Summary ──────────────────────────────────────────────────────────

print(f"\n{_passed} passed, {_failed} failed")
if _failed:
    sys.exit(1)
